plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
-    switch' = fmap not . switch
-
-pPirOpts :: Parser Command
-pPirOpts = hsubparser $
-    command "analyse"
-        (info (Analyse <$> ioSpec) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise and analyse the program, " <>
-              "looking for variables with the largest retained size.")
-  <> command "compile"
-        (info (Compile <$> pCOpts) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise it, " <>
-              "and test if it can be successfully compiled to PLC.")
-  <> command "print"
-        (info (Print <$> printOpts) $
-            progDesc $
-              "Given a PIR program in flat format, " <>
-              "deserialise it and print it out textually.")
-
--- | Load flat pir and deserialize it
-loadPir :: Input -> IO (PirProg ())
-loadPir = loadASTfromFlat Named
-
-compile :: COpts -> PirProg () -> Either PIRError PLCTerm
-compile opts (PIR.Program _ pirT) = do
+
+type PirError a = PIR.Error PLC.DefaultUni PLC.DefaultFun a
+type UnitProvenance = PIR.Provenance ()
+
+---------------- Types for command line options ----------------
+
+-- | A specialised format type for PIR. We don't support deBruijn or named deBruijn for PIR.
+data PirFormat = TextualPir | FlatNamed
+instance Show PirFormat
+    where show = \case { TextualPir  -> "textual"; FlatNamed -> "flat-named" }
+
+data PirOptimiseOptions = PirOptimiseOptions Input PirFormat Output PirFormat PrintMode
+
+data AnalyseOptions = AnalyseOptions Input PirFormat Output -- Input is a program, output is text
+
+data Language = PLC | UPLC
+instance Show Language
+    where show = \case { PLC  -> "plc"; UPLC -> "uplc" }
+
+-- | Compilation options: target language, whether to optimise or not, input and output streams and types
+data CompileOptions =
+    CompileOptions Language
+                   Bool   -- Optimise or not?
+                   Bool   -- True -> just report if compilation was successful; False -> write output
+                   Input
+                   PirFormat
+                   Output
+                   Format
+                   PrintMode
+
+data Command = Analyse  AnalyseOptions
+             | Compile  CompileOptions
+             | Convert  ConvertOptions
+             | Optimise PirOptimiseOptions
+             | Print    PrintOptions
+
+
+---------------- Option parsers ----------------
+
+-- | Invert a switch: return False if the switch is supplied on the command
+-- line, True otherwise.  This is used for command-line options to turn things
+-- off.
+switch' :: Mod FlagFields Bool -> Parser Bool
+switch' = fmap not . switch
+
+pirFormatHelp :: String
+pirFormatHelp =
+  "textual or flat-named (names)"
+
+pirFormatReader :: String -> Maybe PirFormat
+pirFormatReader =
+    \case
+         "textual"    -> Just TextualPir
+         "flat-named" -> Just FlatNamed
+         "flat"       -> Just FlatNamed
+         _            -> Nothing
```

Comment:
\<bikeshedding>
These boilerplate readers (needed for optparse-applicative) could perhaps be removed via some TemplateHaskell. Sth like`$(deriveViaFlipShow ''PirFormat)`

I wonder if somebody has written sth like this before
\</bikeshedding>

Reply:
can we just reuse [Format](https://github.com/input-output-hk/plutus/blob/3bffea1f8738077874dcb859997b9f2970a9be9a/plutus-core/executables/src/PlutusCore/Executable/Common.hs#L310)?

Reply:
Hmmm....I guess the [formatHelp](https://github.com/input-output-hk/plutus/blob/3bffea1f8738077874dcb859997b9f2970a9be9a/plutus-core/executables/src/PlutusCore/Executable/Parsers.hs#L48) would be not right in this case too because the debruijn formats aren't supported.

Reply:
If we want something cheap and auto-derived why not just use [`auto`](https://hackage.haskell.org/package/optparse-applicative-0.17.0.0/docs/Options-Applicative.html#v:auto)?

Reply:
Because it's been like that forever and it's easier to type `flat-named` than `FlatNamed`.